status/templatetags/status_filters.py

```diff
@@ -0,0 +1,170 @@
+    if user.is_anonymous():
+        user = User()  # create a temporary user instance with on attribute
+    return user
+
+
+def show_submission(submission, user):
+    '''Test if the user can see that submission
+
+    Args:
+        submission: a Submission object
+        user: an User object
+    Returns:
+        a boolean of the judgement
+    '''
+    # admin can see all submissions
+    if user.user_level == user.ADMIN:
+        return True
+
+    # no one can see admin's submissions
+    if submission.user.user_level == user.ADMIN:
+        return False
+
+    # user's own submission must be seen
+    if submission.user == user:
+        return True
+
+    # problem owner can see all submission of his problem
+    if submission.problem.owner_id == user.username:
+        return True
+
+    # contest owner & coowner's submission can't be seen until the end
+    contests = Contest.objects.filter(
+        is_homework=False,
+        problem=submission.problem,
+        end_time__gt=datetime.now())
+    if contests:
+        owners = []
+        for contest in contests:
+            owners.append(contest.owner)
+            owners.extend(contest.coowner.all())
+        if user in owners:
+            # owner/coowner can see submission
+            return True
+        else:
+            # not a owner/coowner
+            # to see submission, submission.user must not be owners
+            return submission.user not in owners
+
+    valid = True
+    # an invisible problem's submission can't be seen
+    if not submission.problem.visible:
+        valid = False
+    # problem owner's submission can't be seen
+    if submission.user.username == submission.problem.owner_id:
+        valid = False
+    return valid
+
+
+@register.filter()
+def show_detail(submission, user):
+    '''Test if the user can see that submission's
+    details (code, error message, etc)
+
+    Args:
+        submission: the submission to show
+        user: an User object
+    Returns:
+        a boolean of the judgement
+    '''
+    user = validate_user(user)
+
+    # basic requirement: submission must be shown
+    if show_submission(submission, user):
+        # admin can see everyone's detail
+        if user.user_level == user.ADMIN:
+            return True
+        # no one can see admin's detail
+        if submission.user.user_level == user.ADMIN:
+            return False
+
+        contests = Contest.objects.filter(
+            is_homework=False,
+            start_time__lt=datetime.now(),
+            end_time__gt=datetime.now())
+        # during the contest, only owner/coowner with user level sub-judge/judge 
+        # can view the detail
+        if contests:
+            contests = contests.filter(problem=submission.problem)
+            owners = []
+            for contest in contests:
+                owners.append(contest.owner)
+                owners.extend(contest.coowner.all())
+            if user.has_subjudge_auth() and user in owners:
```

Comment:
An owner or coowner must be sub-judge or judge or admin, so you don't have to check it.